csi: Implement NodeServiceCapability_RPC_GET_VOLUME_STATS rpc call

humblec opened the pull request:
Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Expose CSI volume stats via kubelet volume metrics
```

humblec commented:
/retest

humblec commented:
/assign @gnufied

humblec commented:
/release-note-none

humblec commented:
@gnufied Can you please take a look at this PR ? :)

humblec commented:
/assign @rootfs

shtripat commented on pkg/volume/csi/csi_metrics.go:
```diff
@@ -0,0 +1,119 @@
+
+	if mc.volumeID == "" {
+		return nil, fmt.Errorf("VolumeID is nil")
+
```
Remove the blank line

msau42 commented on pkg/volume/csi/csi_metrics.go:
```diff
@@ -0,0 +1,119 @@
+	driverName csiDriverName
+	specName   string
+	volumeID   string
+	readOnly   bool
```
can you cleanup the unused variables

humblec replied:
@msau42 done. thanks !

yuxiangqian commented on pkg/volume/csi/csi_client.go:
```diff
@@ -111,6 +119,12 @@ const (
 	steps           = 5
 )
 
+type usageCount struct {
```
I am new to K8s community, just out of my curiosity. What are the reasons to introduce a new type here instead of re-use volume.Metrics type? it needs to eventually translated later anyways. Also, is unit testing module required in this particular repo? thanks

humblec replied:
Basically, some drivers return only `BYTES` and some  returns only `INODES`.  A clear separation on these `index` is the main reason. Also, with usageCount driver can set these fields in `int64`.

gnufied replied:
thinking bit back, I think agree with @yuxiangqian. We are doing same `switch..case` in two places which seems wasteful.

humblec replied:
Sure, if more than one vote, it has to be accommodated :) , used volume.Metrics struct in the latest patch set.

humblec commented on pkg/volume/csi/csi_client.go:
```diff
@@ -841,3 +855,109 @@ func (c *csiClientGetter) Get() (csiClient, error) {
+		return c.nodeSupportsVolumeStatsV1(ctx)
+
+	}
+	return false, fmt.Errorf("failed to call NodeSupportsVolumeStats. Both nodeV1ClientCreator and nodeV0ClientCreator are nil")
```
Yeah, lets keep v1 , I changed the error message.

humblec commented on pkg/volume/csi/csi_client.go:
```diff
@@ -841,3 +855,109 @@ func (c *csiClientGetter) Get() (csiClient, error) {
+		return c.nodeGetVolumeStatsV1(ctx, volID, targetPath)
+	}
+
+	return nil, fmt.Errorf("failed to call NodeGetVolumeStats. Both nodeV1ClientCreator and nodeV0ClientCreator are nil")
```
Corrected.

msau42 commented:
/retest

gnufied commented:
@humblec the return value from CSI method looks good. Can you please update the PR for failing tests? We are tracking this for 1.15

humblec commented:
/retest

humblec commented:
@gnufied thanks for the review!!  The tests are passing now,  lets get this in 1.15. :+1:

humblec commented:
/retest

humblec commented:
/retest

gnufied commented on pkg/volume/csi/csi_client.go:
```diff
@@ -72,9 +72,17 @@ type csiClient interface {
+		volID string,
+		targetPath string,
+	) (map[string]usageCount,
+		error)
```
I do not think we need to split the return values on 2 lines.

humblec replied:
Done.

gnufied commented on pkg/volume/csi/csi_client.go:
```diff
@@ -841,3 +855,109 @@ func (c *csiClientGetter) Get() (csiClient, error) {
+		case "UNKNOWN":
+		}
+
+	}
```
Can we add some unit tests for this please?

gnufied commented on pkg/volume/csi/csi_client.go:
```diff
@@ -841,3 +855,109 @@ func (c *csiClientGetter) Get() (csiClient, error) {
+			uci.total = usage.GetTotal()
+			uci.used = usage.GetUsed()
+			usageMap["INODES"] = uci
+		case "UNKNOWN":
```
rather than using `UNKNOWN` here - can we use default case and log the values at error level?

humblec replied:
Done.

gnufied commented on pkg/volume/csi/csi_metrics.go:
```diff
@@ -0,0 +1,108 @@
+		return err
+	}
+	if !volumeStatsSet {
+		klog.Infof(log("metricsCsi.getCSIVolInfo GET_VOLUME_STATS capability not set. Skipping NodeVolumeStats..."))
```
If the caller of `GetMetrics` logs the errors we should not log these errors here again.

humblec replied:
These logs are just meant for mentioning where exactly we failed, seems ok to me, but if you think this is really unnecessary please let me know.

gnufied replied:
yeah I am afraid, these lines will be too noisy in production environments. It is not a hard error if CSI  volume does not implement this RPC call.

gnufied replied:
not removed yet.

humblec replied:
Done, in latest patchset. Sorry for missing it @gnufied in previous one.

gnufied commented on pkg/volume/csi/csi_metrics.go:
```diff
@@ -0,0 +1,108 @@
+			metrics.Inodes = resource.NewQuantity(v.total, resource.BinarySI)
+		} else {
+			continue
+		}
```
unnecessary continue?

humblec replied:
Removed.

gnufied commented on pkg/volume/csi/csi_metrics.go:
```diff
@@ -0,0 +1,108 @@
+	return &metricsCsi{volumeID: volumeID, targetPath: targetPath}
+}
+
+func (mc *metricsCsi) GetMetrics() (*volume.Metrics, error) {
```
Can we add some unit tests please?

humblec replied:
Done.

gnufied commented:
It would be also nice to have some e2e that use CSIMock driver to return some actual stats and verify they are available from kubelet. We already have plenty of tests that use CSIMock driver. You can probably use same test harness to add a new test.

humblec commented:
/retest

gnufied commented on pkg/volume/csi/csi_metrics_test.go:
```diff
@@ -0,0 +1,119 @@
+	}{
+
+		{
+			name:           "when nodeVolumeStats=on, DevicePath=on, DeviceMountPath=on",
```
lets remove `devicePath` from these names?

humblec replied:
Sure, replaced `devicePath` with `VolumeID`

gnufied commented on pkg/volume/csi/csi_client.go:
```diff
@@ -841,3 +854,110 @@ func (c *csiClientGetter) Get() (csiClient, error) {
+	for _, capability := range capabilities {
+		if capability.GetRpc().GetType() == csipbv1.NodeServiceCapability_RPC_GET_VOLUME_STATS {
+			volumeStatsSet = true
+		}
```
might as well get rid of `volumeStatsSet` variable and just return true and false from whenever applicable.

humblec replied:
Done. Removed `volumeStatsSet` var.

gnufied commented on pkg/volume/csi/csi_metrics_test.go:
```diff
@@ -0,0 +1,119 @@
+
+		csClient := setupClientWithVolumeStats(t, tc.volumeStatsSet)
+
+		_, err := csClient.NodeGetVolumeStats(ctx, csiSource.VolumeHandle, tc.volumeData.DeviceMountPath)
```
If this function just tests `NodeGetVolumeStats`, it should be kept with other similar tests in csi_client_test.go. I was hoping that, we will have a higher level of csi metrics tests here and we won't need csi client tests then.

humblec replied:
Kept it in a different file, so that we can collect/capture metrics related tests ( in future) in a single place. Thanks

gnufied replied:
I still think this test has been added to incorrect file. All tests that check `csiClient` interface belong to `csi_client_test.go` - I am not sure why this should be an exception.

This test file should instead be testing `GetMetrics` function defined in `MetricsProvider` interface. We also need actual parsing of stats info to be tested, which can be done by setting response of `NodeGetVolumeStats` in `NodeClient` object. There is already a example of it where we set `nodeGetInfoResp` in `NodeClient`.

humblec replied:
>I still think this test has been added to incorrect file. All tests that check csiClient interface belong to csi_client_test.go
>
Done, moved it to csi_client_test.go

gnufied reviewed (commented):
Looking pretty close to done. We just need to clean up the tests. I am fine adding e2e as a follow up.

humblec commented:
/retest

humblec commented:
/retest

gnufied commented on pkg/volume/csi/csi_client.go:
```diff
@@ -841,3 +854,109 @@ func (c *csiClientGetter) Get() (csiClient, error) {
+	}
+	for _, usage := range usages {
+
+		unit = usage.GetUnit()
```
oddly place new line.

humblec replied:
Done.

gnufied commented on pkg/volume/csi/csi_client.go:
```diff
@@ -841,3 +854,109 @@ func (c *csiClientGetter) Get() (csiClient, error) {
+		return nil, err
+	}
+
+	usages := resp.GetUsage()
```
oddly placed new line.

humblec replied:
Done.

gnufied commented on pkg/volume/csi/csi_metrics_test.go:
```diff
@@ -0,0 +1,119 @@
+		defer cancel()
+
+		csiSource, _ := getCSISourceFromSpec(tc.volumeData.VolumeSpec)
+
```
I would remove these unnecessary new lines.

humblec replied:
Done.

gnufied commented on pkg/volume/csi/csi_metrics_test.go:
```diff
@@ -0,0 +1,119 @@
+		_, err := csClient.NodeGetVolumeStats(ctx, csiSource.VolumeHandle, tc.volumeData.DeviceMountPath)
+
+		if err != nil && tc.success {
+			t.Errorf("For %s : expected %v got %v", tc.name, tc.success, err)
```
None of the tests exercise if parsing of `INODES` and `BYTES` from grpc response work. Can we fix that?

gnufied replied:
Can we fix this?

gnufied commented on pkg/volume/csi/csi_client.go:
```diff
@@ -841,3 +854,109 @@ func (c *csiClientGetter) Get() (csiClient, error) {
+			uci.total = usage.GetTotal()
+			uci.used = usage.GetUsed()
+			usageMap["INODES"] = uci
+		case "default":
```
I meant `default:` case, not literally "default" string. It is unlikely we will get "default" string there.

humblec replied:
ah.. zombie eyes , corrected it.

humblec commented:
/retest

humblec commented:
@gnufied addressed the comments. PTAL

gnufied commented on pkg/volume/csi/csi_metrics.go:
```diff
@@ -0,0 +1,94 @@
+
+	if mc.targetPath == "" {
+		return nil, fmt.Errorf("targetpath is nil")
+	}
```
Aren't we already testing these fields in `NodGetVolumeStats` function -why do we need to check here again?

humblec replied:
It was an extra check for fail early. Removed now as per above comment.

gnufied commented on pkg/volume/csi/csi_metrics.go:
```diff
@@ -0,0 +1,94 @@
+	return metrics, nil
+}
+
+func (mc *metricsCsi) getCSIVolInfo(metrics *volume.Metrics) error {
```
Do we need a different function here? We can merge this inside `GetMetrics` function itself.

humblec replied:
Merged getCSIVolInfo to GetMetrics. Done.

humblec commented:
/retest

humblec commented:
/retest

humblec commented:
/retest

humblec commented:
@gnufied enahanced the PR with tests . I believe all the comments are addressed. PTAL .. Thanks a lot!!

gnufied commented on pkg/volume/csi/csi_metrics_test.go:
```diff
@@ -0,0 +1,114 @@
+				availableBytes := resource.NewQuantity(expectedMetrics.Usage[i].Available, resource.BinarySI)
+				totalBytes := resource.NewQuantity(expectedMetrics.Usage[i].Total, resource.BinarySI)
+				usedBytes := resource.NewQuantity(expectedMetrics.Usage[i].Used, resource.BinarySI)
+				if *(metrics.Available) != *availableBytes {
```
For `resource.Quantity` it might be better to use `metrics.Available.Cmp(availableBytes)`.

humblec replied:
Sure, changed to `Cmp` .

gnufied commented on pkg/volume/csi/csi_metrics_test.go:
```diff
@@ -0,0 +1,114 @@
+		for i, usage := range expectedMetrics.Usage {
+			if usage.Unit == csipbv1.VolumeUsage_BYTES {
+				availableBytes := resource.NewQuantity(expectedMetrics.Usage[i].Available, resource.BinarySI)
+				totalBytes := resource.NewQuantity(expectedMetrics.Usage[i].Total, resource.BinarySI)
```
can we replace `expectedMetrics.Usage[i].Available` with `usage.Available` ?

humblec replied:
Done.

gnufied commented on pkg/volume/csi/csi_metrics_test.go:
```diff
@@ -0,0 +1,114 @@
+				freeInodes := resource.NewQuantity(expectedMetrics.Usage[i].Available, resource.BinarySI)
+				inodes := resource.NewQuantity(expectedMetrics.Usage[i].Total, resource.BinarySI)
+				usedInodes := resource.NewQuantity(expectedMetrics.Usage[i].Used, resource.BinarySI)
+				if *(metrics.InodesFree) != *freeInodes {
```
Same as above.

humblec replied:
Done.

gnufied commented on pkg/volume/csi/csi_client.go:
```diff
@@ -841,3 +849,97 @@ func (c *csiClientGetter) Get() (csiClient, error) {
+		return nil, err
+	}
+	usages := resp.GetUsage()
+	fmt.Printf("metrics is %+v", usages)
```
lets remove this debug `fmt.Printf` ?

humblec replied:
Sure, removed.

gnufied commented on pkg/volume/csi/csi_client.go:
```diff
@@ -841,3 +849,97 @@ func (c *csiClientGetter) Get() (csiClient, error) {
+	metrics := &volume.Metrics{}
+	if usages == nil {
+		return nil, nil
+	}
```
if `usages` is nil there is no point in initializing `metrics` variable. So, I would move `if usages == nil` check before that declaration.

gnufied replied:
Also - IMO we should return an error if `usages` is nil because if a plugin implements this RPC call and still returns no usage information that most likely means something is incorrect.

humblec replied:
Addressed above comments.

gnufied commented on pkg/volume/csi/csi_client.go:
```diff
@@ -841,3 +849,97 @@ func (c *csiClientGetter) Get() (csiClient, error) {
+	targetPath string,
+) (*volume.Metrics, error) {
+
+	var unit csipbv1.VolumeUsage_Unit
```
Can we remove this variable from here? We could always do `unit := ` in the loop below.

humblec replied:
Done.

gnufied commented on pkg/volume/csi/csi_client_test.go:
```diff
@@ -61,6 +69,58 @@ func (c *fakeCsiDriverClient) NodeGetInfo(ctx context.Context) (
+
+	c.t.Logf("c:%+v and nodeclient:%+v", c, c.nodeClient)
+	resp, err := c.nodeClient.NodeGetVolumeStats(ctx, req)
+	c.t.Logf("Response: %v", resp)
```
We should not log full object in these tests. In CI logs, they will generate unnecessary noisy output.

humblec replied:
Removed.

gnufied commented on pkg/volume/csi/csi_metrics.go:
```diff
@@ -0,0 +1,77 @@
+	}
+	if metrics == nil {
+		return nil, fmt.Errorf("metrics is nil")
+	}
```
Little more information in the error message perhaps. Something like `fmt.Errorf("csi.NodeGetVolumeStats returned nil metrics for volume %s", mc.volumeID")`

humblec replied:
Converted to above.

gnufied commented on pkg/volume/csi/csi_metrics_test.go:
```diff
@@ -0,0 +1,114 @@
+)
+
+func TestGetMetrics(t *testing.T) {
+
```
unncessary new line.

humblec replied:
Removed.

gnufied commented:
/kind feature
/priority important-soon

gnufied commented:
Also @humblec can you add a release-note to the PR? Something like - "Expose CSI volume stats via kubelet volume metrics" ?

humblec commented:
>Also @humblec can you add a release-note to the PR? Something like - "Expose CSI volume stats via kubelet volume metrics" ?
>
Done. ! 
Addressed the comments too. Thanks @gnufied

gnufied commented on pkg/volume/csi/csi_metrics_test.go:
```diff
@@ -0,0 +1,113 @@
+				availableBytes := resource.NewQuantity(usage.Available, resource.BinarySI)
+				totalBytes := resource.NewQuantity(usage.Total, resource.BinarySI)
+				usedBytes := resource.NewQuantity(usage.Used, resource.BinarySI)
+				if (*(metrics.Available)).Cmp(*availableBytes) != 0 {
```
tiny nit - can we write this compare like `metrics.Available.Cmp(*availableBytes)` ?

humblec replied:
Done.

gnufied commented:
One tiny nit and please squash all the commits in one.  rest looks good to me.

gnufied commented on pkg/volume/csi/csi_metrics_test.go:
```diff
@@ -0,0 +1,113 @@
+				if (*(metrics.Inodes)).Cmp(*inodes) != 0 {
+					t.Fatalf("for %s: error: expected :%v , got: %v", tc.name, *inodes, *(metrics.Inodes))
+				}
+				if (*(metrics.InodesUsed)).Cmp(*usedInodes) != 0 {
```
same here about use of `Cmp` function.

humblec replied:
Done.

humblec committed:
csi: Implement NodeServiceCapability_RPC_GET_VOLUME_STATS rpc call
and implement Metrics Provider for CSI driver

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
>One tiny nit and please squash all the commits in one. rest looks good to me.
>
Sure, its done in latest patch set.. Thanks @gnufied  !!

gnufied commented:
/lgtm

gnufied commented:
/assign @msau42

msau42 commented on pkg/volume/csi/csi_client.go:
```diff
@@ -841,3 +849,94 @@ func (c *csiClientGetter) Get() (csiClient, error) {
+			metrics.Inodes = resource.NewQuantity(usage.GetTotal(), resource.BinarySI)
+			metrics.InodesUsed = resource.NewQuantity(usage.GetUsed(), resource.BinarySI)
+		default:
+			klog.Errorf("unknown key %s in usage", unit.String())
```
This means that CSI spec added more that we don't support yet?

Should it be a warning, or even Info at level 5 if we expect this to flood logs?

gnufied replied:
This should ideally never happen and if it does, it is probably an error and worth either fixing k8s or CSI driver.

msau42 commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -408,6 +408,8 @@ func (p *csiPlugin) NewMounter(
 	}
 	klog.V(4).Info(log("created path successfully [%s]", dataDir))
 
+	mounter.MetricsProvider = NewMetricsCsi(volumeHandle, dataDir)
```
Should the path be dir or dataDir?  dir is the one we pass as the targetPath to NodePublish?

gnufied replied:
good catch, this indeed should be `dir` and not `dataDir`.

msau42 replied:
Can we make sure to validate this with a real CSI driver?

humblec replied:
Replaced the path. 
>Can we make sure to validate this with a real CSI driver?
>
Yes, thats the very next plan. I dont think, the CSI drivers exist which expose the metrics. I have started the work for Ceph-CSI any way.

gnufied replied:
This can be also tested via CSI mock driver fwiw. Mock driver can implement `NodeGetVolumeStats` RPC call and it can check for volume path to be same as publish path and return error if it isn't.  

This has to go in an e2e though. @humblec can you add a github issue for creating an e2e for this?

humblec replied:
Sure @gnufied . I am opening an issue for e2e tracking as discussed earlier..

humblec replied:
@gnufied The E2E issue is here https://github.com/kubernetes/kubernetes/issues/77933 . Thanks !

gnufied commented on pkg/volume/csi/csi_client.go:
```diff
@@ -841,3 +849,94 @@ func (c *csiClientGetter) Get() (csiClient, error) {
+}
+
+func (c *csiDriverClient) NodeGetVolumeStats(ctx context.Context, volID string, targetPath string) (*volume.Metrics, error) {
+	klog.V(4).Info(log("calling NodeGetVolumeStats rpc: [volid=%s, target_path=%s", volID, targetPath))
```
As often as prometheus is configured to scrap the metrics endpoint.

Davidrjx replied:
unable to scrap any metric for aws efs csi , not sure whether rpc call was executed or not.

humblec commented on pkg/volume/csi/csi_client.go:
```diff
@@ -841,3 +849,94 @@ func (c *csiClientGetter) Get() (csiClient, error) {
 }
+
+func (c *csiDriverClient) NodeSupportsVolumeStats(ctx context.Context) (bool, error) {
+	klog.V(4).Info(log("calling NodeGetCapabilities rpc to determine if NodeSupportsVolumeStats"))
```
lifted to V(5).

humblec commented on pkg/volume/csi/csi_client.go:
```diff
@@ -841,3 +849,94 @@ func (c *csiClientGetter) Get() (csiClient, error) {
+	for _, usage := range usages {
+		unit := usage.GetUnit()
+		switch unit.String() {
+		case "BYTES":
```
Converted to consts.

humblec commented:
@msau42 the final review comments are addressed. PTAL .. Thanks !!

gnufied commented:
@humblec this PR was discussed in wg-csi-implementation stadnup and it was agreed that we should try and manually validate if CSI volume stats are being available through kubelet metrics if using a real CSI driver (even mock or hostpath driver would do). 

We can do e2e as a follow up but could we do the manual validation for ANY CSI driver?

msau42 commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -408,6 +408,8 @@ func (p *csiPlugin) NewMounter(
 	}
 	klog.V(4).Info(log("created path successfully [%s]", dataDir))
 
+	mounter.MetricsProvider = NewMetricsCsi(volumeHandle, dir)
```
This still doesn't look like the right directory?

msau42 replied:
wait no i can't read this looks right

msau42 commented:
/approve
/hold 
for testing with a real csi driver

humblec commented:
>@humblec this PR was discussed in wg-csi-implementation stadnup and it was agreed that we should try and manually validate if CSI volume stats are being available through kubelet metrics if using a real CSI driver (even mock or hostpath driver would do).

We can do e2e as a follow up but could we do the manual validation for ANY CSI driver?
>

Sure @gnufied and @msau42 . I patched hostpath driver to return a handcrafted response for this RPC call and tested out.  The final result is that, I could see the metrics in kubelet. Below is an example output for the RPC call and also the metrics in kubelet.


Logs from CSI driver:

```
I0516 15:55:17.079907       1 server.go:117] GRPC call: /csi.v1.Node/NodeGetVolumeStats
I0516 15:55:17.079920       1 server.go:118] GRPC request: {"volume_id":"c0419fd7-77f2-11e9-b47e-c85b7636c232","volume_path":"/var/lib/kubelet/pods/4891f5a7-2c72-4f33-8e4b-88ba63b5eb1a/volumes/kubernetes.io~csi/pvc-b1e01c02-063a-467c-9f88-621cadbf780f/mount"}
I0516 15:55:17.080500       1 mount_linux.go:164] Detected OS without systemd
I0516 15:55:17.080510       1 nodeserver.go:302] VolumeStats: stats on targetpath/volumeID /var/lib/kubelet/pods/4891f5a7-2c72-4f33-8e4b-88ba63b5eb1a/volumes/kubernetes.io~csi/pvc-b1e01c02-063a-467c-9f88-621cadbf780f/mount/c0419fd7-77f2-11e9-b47e-c85b7636c232 has been requested.
E0516 15:55:17.080519       1 nodeserver.go:304] VolumeStats: stats on targetpath/volumeID /var/lib/kubelet/pods/4891f5a7-2c72-4f33-8e4b-88ba63b5eb1a/volumes/kubernetes.io~csi/pvc-b1e01c02-063a-467c-9f88-621cadbf780f/mount/c0419fd7-77f2-11e9-b47e-c85b7636c232 has been requested.
I0516 15:55:17.080784       1 nodeserver.go:316] targetpath /var/lib/kubelet/pods/4891f5a7-2c72-4f33-8e4b-88ba63b5eb1a/volumes/kubernetes.io~csi/pvc-b1e01c02-063a-467c-9f88-621cadbf780f/mount: already mounted
I0516 15:55:17.080794       1 server.go:123] GRPC response: {"usage":[{"total":10,"unit":1}]}

```

Logs/stats colllected in kubelet!

```

kubelet_volume_stats_available_bytes{namespace="default",persistentvolumeclaim="csi-pvc"} 0
kubelet_volume_stats_capacity_bytes{namespace="default",persistentvolumeclaim="csi-pvc"} 10
kubelet_volume_stats_inodes{namespace="default",persistentvolumeclaim="csi-pvc"} 0
kubelet_volume_stats_inodes_free{namespace="default",persistentvolumeclaim="csi-pvc"} 0
kubelet_volume_stats_inodes_used{namespace="default",persistentvolumeclaim="csi-pvc"} 0
kubelet_volume_stats_used_bytes{namespace="default",persistentvolumeclaim="csi-pvc"} 0

```

Hope this is good to go in and declare the metrics support in CSI !!..

gnufied commented:
/hold cancel
/lgtm

Davidrjx commented:
> > @humblec this PR was discussed in wg-csi-implementation stadnup and it was agreed that we should try and manually validate if CSI volume stats are being available through kubelet metrics if using a real CSI driver (even mock or hostpath driver would do).
> > We can do e2e as a follow up but could we do the manual validation for ANY CSI driver?
> 
> Sure @gnufied and @msau42 . I patched hostpath driver to return a handcrafted response for this RPC call ( also note/confirm the path `/var/lib/kubelet/pods/4891f5a7-2c72-4f33-8e4b-88ba63b5eb1a/volumes/kubernetes.io~csi/pvc-b1e01c02-063a-467c-9f88-621cadbf780f/mount"` in the request) and tested out. The final result is that, I could see the metrics in kubelet. Below is an example output for the RPC call and also the metrics in kubelet.
> 
> Logs from CSI driver:
> 
> ```
> I0516 15:55:17.079907       1 server.go:117] GRPC call: /csi.v1.Node/NodeGetVolumeStats
> I0516 15:55:17.079920       1 server.go:118] GRPC request: {"volume_id":"c0419fd7-77f2-11e9-b47e-c85b7636c232","volume_path":"/var/lib/kubelet/pods/4891f5a7-2c72-4f33-8e4b-88ba63b5eb1a/volumes/kubernetes.io~csi/pvc-b1e01c02-063a-467c-9f88-621cadbf780f/mount"}
> I0516 15:55:17.080500       1 mount_linux.go:164] Detected OS without systemd
> I0516 15:55:17.080510       1 nodeserver.go:302] VolumeStats: stats on targetpath/volumeID /var/lib/kubelet/pods/4891f5a7-2c72-4f33-8e4b-88ba63b5eb1a/volumes/kubernetes.io~csi/pvc-b1e01c02-063a-467c-9f88-621cadbf780f/mount/c0419fd7-77f2-11e9-b47e-c85b7636c232 has been requested.
> E0516 15:55:17.080519       1 nodeserver.go:304] VolumeStats: stats on targetpath/volumeID /var/lib/kubelet/pods/4891f5a7-2c72-4f33-8e4b-88ba63b5eb1a/volumes/kubernetes.io~csi/pvc-b1e01c02-063a-467c-9f88-621cadbf780f/mount/c0419fd7-77f2-11e9-b47e-c85b7636c232 has been requested.
> I0516 15:55:17.080784       1 nodeserver.go:316] targetpath /var/lib/kubelet/pods/4891f5a7-2c72-4f33-8e4b-88ba63b5eb1a/volumes/kubernetes.io~csi/pvc-b1e01c02-063a-467c-9f88-621cadbf780f/mount: already mounted
> I0516 15:55:17.080794       1 server.go:123] GRPC response: {"usage":[{"total":10,"unit":1}]}
> ```
> 
> Logs/stats colllected in kubelet!
> 
> ```
> 
> kubelet_volume_stats_available_bytes{namespace="default",persistentvolumeclaim="csi-pvc"} 0
> kubelet_volume_stats_capacity_bytes{namespace="default",persistentvolumeclaim="csi-pvc"} 10
> kubelet_volume_stats_inodes{namespace="default",persistentvolumeclaim="csi-pvc"} 0
> kubelet_volume_stats_inodes_free{namespace="default",persistentvolumeclaim="csi-pvc"} 0
> kubelet_volume_stats_inodes_used{namespace="default",persistentvolumeclaim="csi-pvc"} 0
> kubelet_volume_stats_used_bytes{namespace="default",persistentvolumeclaim="csi-pvc"} 0
> ```
> 
> Hope this is good to go in and declare the metrics support in CSI !!.. Thanks a lot!

hi @humblec 
did you hard-coded specific pv in driver and execute GetVolumeStats rpc call?
